### ISSUE TITLE
Training Breakout - low performance using default parameters

### ISSUE THREAD

**Issue by allengee** [keras-rl/keras-rl#26]
I trained a network to play Breakout using your default parameters:
python dqn_atari.py

However, the resulting network did not have good performance.  During testing, the paddle stayed mostly in the right corner with only occasional moves towards the center.  The scores were very low.  I assume your pretrained weights for Breakout have good perfomance.  Did you use the same network structure and number of training steps as the default settings?  If not, what settings should I use to effectively train on Breakout?  

Also, I am using TensorFlow with a Titan X Pascal GPU.  The training time shown below are only about 20% faster than training using just a CPU.  In this situation the GPU is not giving the expected 10-100x speed boost vs the CPU.  Is this because the agent/network processing time is only a small fraction of the whole reinforcement learning cycle?  Can you verify that the training times below are about what you would expect using a GPU?  

Thanks,
Allen

Training for 1750000 steps ...
Interval 1 (0 steps performed)
10000/10000 [==============================] - 136s - reward: 0.0062  
Interval 2 (10000 steps performed)
10000/10000 [==============================] - 102s - reward: 0.0058  
Interval 3 (20000 steps performed)
10000/10000 [==============================] - 95s - reward: 0.0062  
Interval 4 (30000 steps performed)
10000/10000 [==============================] - 98s - reward: 0.0061  
Interval 5 (40000 steps performed)
10000/10000 [==============================] - 102s - reward: 0.0064  
Interval 6 (50000 steps performed)
10000/10000 [==============================] - 235s - reward: 0.0057 - loss: 0.0010 - mean_absolute_error: 0.0019 - mean_q: 0.0130 - mean_eps: 0.9522  
Interval 7 (60000 steps performed)
10000/10000 [==============================] - 236s - reward: 0.0058 - loss: 0.0011 - mean_absolute_error: 0.0020 - mean_q: 0.0146 - mean_eps: 0.9432  
Interval 8 (70000 steps performed)
10000/10000 [==============================] - 241s - reward: 0.0064 - loss: 0.0010 - mean_absolute_error: 0.0019 - mean_q: 0.0215 - mean_eps: 0.9342  
Interval 9 (80000 steps performed)
10000/10000 [==============================] - 243s - reward: 0.0056 - loss: 0.0010 - mean_absolute_error: 0.0019 - mean_q: 0.0265 - mean_eps: 0.9252  
Interval 10 (90000 steps performed)
10000/10000 [==============================] - 241s - reward: 0.0056 - loss: 9.9363e-04 - mean_absolute_error: 0.0018 - mean_q: 0.0326 - mean_eps: 0.9162         

...

Interval 175 (1740000 steps performed)
10000/10000 [==============================] - 244s - reward: 0.0187 - loss: 0.0035 - mean_absolute_error: 0.0059 - mean_q: 0.4626 - mean_eps: 0.1000  
done, took 41743.480 seconds
Episode 1: reward: 1.000, steps: 53
Episode 2: reward: 1.000, steps: 53
Episode 3: reward: 1.000, steps: 53
Episode 4: reward: 1.000, steps: 53
Episode 5: reward: 1.000, steps: 53
Episode 6: reward: 1.000, steps: 53
Episode 7: reward: 1.000, steps: 53
Episode 8: reward: 1.000, steps: 53
Episode 9: reward: 1.000, steps: 53
Episode 10: reward: 1.000, steps: 53

**Comment by matthiasplappert** [comment]
Yes, I have used these exact parameter configurations to obtain the results that you can see in the GIF from the readme file. However, I have trained my model using the Theano backend, so that might have something to do with it.

Regarding the performance problems: Generally speaking, you won't get as much out of the GPU as you would with other tasks like training a model on ImageNet. This is because the environment (in this case the Atari simulator) needs to be running as well and the GPU is not fully utilized (try running `nvidia-smi` while training to see what I mean). However, on my GTX 980ti using Theano, I get far better performance than you report:

```
Interval 1 (0 steps performed)
10000/10000 [==============================] - 70s - reward: 0.0072
Interval 2 (10000 steps performed)
10000/10000 [==============================] - 55s - reward: 0.0069
Interval 3 (20000 steps performed)
10000/10000 [==============================] - 56s - reward: 0.0055
Interval 4 (30000 steps performed)
10000/10000 [==============================] - 56s - reward: 0.0059
Interval 5 (40000 steps performed)
10000/10000 [==============================] - 56s - reward: 0.0067
Interval 6 (50000 steps performed)
10000/10000 [==============================] - 138s - reward: 0.0059 - loss: 9.1540e-04 - mean_absolute_error: 0.0024 - mean_q: 0.0240 - mean_eps: 0.9522
Interval 7 (60000 steps performed)
10000/10000 [==============================] - 82s - reward: 0.0055 - loss: 5.6205e-04 - mean_absolute_error: 0.0015 - mean_q: 0.0202 - mean_eps: 0.9432
Interval 8 (70000 steps performed)
10000/10000 [==============================] - 85s - reward: 0.0058 - loss: 4.7278e-04 - mean_absolute_error: 0.0019 - mean_q: 0.0243 - mean_eps: 0.9342
```
Notice that the first training time step is slightly longer since the model initially compiles the train function. After that, I'm getting between 80 - 90s per 10,000 steps, which is 3x faster than what you have reported.

Can you try to run the Atari example using Theano backend (also set `image_dim_ordering` to `th` in `~/.keras/keras.json`) and see if that improves things both performance-wise and also with respect to the agent's performance?

**Comment by matthiasplappert** [comment]
Also, I'll let it train for the 1.75M iterations again to see if I can reproduce the problem. In any case, I'll post the training log here afterwards so you can compare it to what you are seeing.

**Comment by matthiasplappert** [comment]
I've successfully trained the `dqn_atari.py` example on master (897f101f4a2420d917c5b81d53fff11b210ae520) using the Theano backend. Here's the result when testing after 1.75M iterations:

```
Episode 1: reward: 45.000, steps: 1151
Episode 2: reward: 25.000, steps: 851
Episode 3: reward: 37.000, steps: 1165
Episode 4: reward: 38.000, steps: 1090
Episode 5: reward: 41.000, steps: 1141
Episode 6: reward: 33.000, steps: 926
Episode 7: reward: 27.000, steps: 889
Episode 8: reward: 43.000, steps: 1146
Episode 9: reward: 28.000, steps: 904
Episode 10: reward: 58.000, steps: 1262
```
I've also update the weights repository with the final set of weights, all snapshots during training and the entire training log. You can find them here: https://github.com/matthiasplappert/keras-rl-weights/tree/master/dqn_atari/breakout.

So basically I do not know why it doesn't work for you. Maybe really give it a try using Theano instead (should also be faster) and see if that works. I'll also try to reproduce the problem when using the TF backend.

**Comment by allengee** [comment]
I am running on Ubuntu 14.04 using the following software versions: 
Keras 1.1.0, Theano 0.8.2 with CUDA 8.0rc and cudnn 5.0, TensorFlow 0.10.0 with CUDA 7.5 and cudnn 5.1, gym 0.2.11
What OS and software versions are you using?

I tested your final pretrained weights using Theano with GPU and got exactly the same output you posted.  I then trained my own weights and got these testing results at the end of the training run:

Interval 175 (1740000 steps performed)
10000/10000 [==============================] - 100s - reward: 0.0296 - loss: 0.0022 - mean_absolute_error: 0.0097 - mean_q: 2.0867 - mean_eps: 0.1000         
done, took 17819.611 seconds
Episode 1: reward: 2.000, steps: 103
Episode 2: reward: 5.000, steps: 195
Episode 3: reward: 9.000, steps: 343
Episode 4: reward: 5.000, steps: 197
Episode 5: reward: 10.000, steps: 391
Episode 6: reward: 12.000, steps: 444
Episode 7: reward: 4.000, steps: 172
Episode 8: reward: 9.000, steps: 327
Episode 9: reward: 23.000, steps: 591
Episode 10: reward: 13.000, steps: 394

But when I loaded the saved weights and retested them, I got:

Episode 1: reward: 30.000, steps: 864
Episode 2: reward: 31.000, steps: 837
Episode 3: reward: 33.000, steps: 1110
Episode 4: reward: 51.000, steps: 1336
Episode 5: reward: 37.000, steps: 1130
Episode 6: reward: 22.000, steps: 747
Episode 7: reward: 38.000, steps: 936
Episode 8: reward: 39.000, steps: 1260
Episode 9: reward: 66.000, steps: 1446
Episode 10: reward: 36.000, steps: 1025

The two testing results above should be about the same since they are supposed to be using the same weights, but they are not! The first results are mediocre, while the second are similar to what you had posted. Do you know why this would be?  
There seems to be a bug here, although one possibility is the game gets hard the longer you play it so that the first results are against a harder game.  Can you reproduce the effect?

The other difference I noticed is that your training times are 15-20% faster than mine. My training interval time is 103s +/- 3s.  This is strange since I built my machine only a few weeks ago and have all the latest hardware.  

Are you doing any optimizations that I may have missed?
My .theanorc file is simply:

[global]
floatX = float32
device = gpu0

[lib]
cnmem = 1

Are you using any other .theanorc settings that might speed things up?

As another speed comparison, can you run the standard test for TensorFlow with GPU?
My results are:

python -m tensorflow.models.image.mnist.convolutional

Step 0 (epoch 0.00), 553.2 ms
Minibatch loss: 12.054, learning rate: 0.010000
Minibatch error: 90.6%
Validation error: 84.6%
Step 100 (epoch 0.12), 7.0 ms
Minibatch loss: 3.283, learning rate: 0.010000
Minibatch error: 6.2%
Validation error: 6.7%
Step 200 (epoch 0.23), 6.4 ms
Minibatch loss: 3.501, learning rate: 0.010000
Minibatch error: 15.6%
Validation error: 3.6%
Step 300 (epoch 0.35), 6.6 ms
Minibatch loss: 3.210, learning rate: 0.010000
Minibatch error: 4.7%
Validation error: 3.4%
Step 400 (epoch 0.47), 6.4 ms
Minibatch loss: 3.226, learning rate: 0.010000
Minibatch error: 10.9%
Validation error: 2.9%
Step 500 (epoch 0.58), 6.4 ms
Minibatch loss: 3.317, learning rate: 0.010000
Minibatch error: 9.4%
Validation error: 2.5%

Thanks,
Allen

**Comment by matthiasplappert** [comment]
My setup uses Ubuntu 14.04 on a Intel i7-6700K with CUDA 7.5 and cuDNN v4, Keras 1.0.7, Theano 0.9rc1 and TensorFlow 0.9.0. My Theano config is similar but I also use the `fastmath` option, which speeds up certain operations for the sake of losing precision (see Theano documentation). My file looks like this:
```
[global]
floatX = float32
device = gpu

[nvcc]
fastmath = True

[lib]
cnmem = 0.8
```
The only other reason I can think of why my setup is faster is that I have installed OpenBLAS (`sudo apt-get install libopenblas-dev`), which is a lot faster than standard BLAS. However, I'm not even sure if it it used at all if the network is run on the GPU.

Regarding the lower reward in after training: This is because during training the game is only run until the first live is lost whereas in test mode the game is run until it is truly game over (see `_step` implementation for details). I agree that this is slightly confusing but explains the different results during train and test time.

Here are the results for TensorFlow on my system:
```
Step 0 (epoch 0.00), 11.0 ms
Minibatch loss: 12.054, learning rate: 0.010000
Minibatch error: 90.6%
Validation error: 84.6%
Step 100 (epoch 0.12), 5.7 ms
Minibatch loss: 3.281, learning rate: 0.010000
Minibatch error: 6.2%
Validation error: 6.9%
Step 200 (epoch 0.23), 5.5 ms
Minibatch loss: 3.501, learning rate: 0.010000
Minibatch error: 14.1%
Validation error: 3.8%
Step 300 (epoch 0.35), 5.5 ms
Minibatch loss: 3.221, learning rate: 0.010000
Minibatch error: 9.4%
Validation error: 3.2%
Step 400 (epoch 0.47), 5.4 ms
Minibatch loss: 3.213, learning rate: 0.010000
Minibatch error: 9.4%
Validation error: 2.8%
Step 500 (epoch 0.58), 5.5 ms
Minibatch loss: 3.288, learning rate: 0.010000
Minibatch error: 7.8%
Validation error: 2.8%
```

**Comment by yongduek** [comment]
Try to use a fractional number of cnmem. I saw an article saying that cnmem has some effect on the performance of then.

For example,

[lib]
cnmem = 0.8

**Comment by allengee** [comment]
I've already switched to using cnmem = 0.8.  But didn't notice any difference in doing so.

Matthias,

I'm using the 3.6GHz 6-core Broadwell-E 6850K CPU. You are using a 4.0GHz 4-core CPU.
In this case having higher CPU clock speed seems more advantageous than having more cores.
I think this explains why your system is a bit faster than mine on this task.
I was also able to get 5-10% performance improvement by optimizing the BIOS settings.
I'm running 2400MHz RAM.  Are you using a higher frequency on yours?

I tried installing OpenBLAS, but it didn't seem to have much effect.  However, the interval training times are very much statistical with a standard deviation of about 3%.  This can be seen by running for 10 intervals, exiting and then running again several times. This makes it hard to measure small effects. 

I think we have verified that my system is working similar to yours with the Theano backend.  I'm still getting the same bad performance with the TensorFlow backend, even though the TensorFlow installation test results look good.  Have you been able to verify that dqn_atari.py produces good results using the TensorFlow backend on your system?

**Comment by matthiasplappert** [comment]
I'm currently working on several improvements in DQN, e.g. getting rid of the extra forward pass during training. I'll test it with TensorFlow once this is done to make sure that it works properly.

**Comment by matthiasplappert** [comment]
I've just merged some pretty significant changes to DQN. Overall, the training should now be much faster. There were also some problems with the way the replay memory handled starts of episodes, so things should be improved here as well.
